gigascience/gigadb-website
# a11y - Article detail page - essential fixes

# Pull request for issue: #1371 

This is a pull request for the following functionalities:

* Improve accessibiltiy of dataset page
  * Maintain consistent heading levels
  * Enhance checkbox with visible focus state
  * Replace empty <section> tags with spacing CSS class
  * Link labels to corresponding inputs
  * Group checkboxes within a fieldset
  * Add legends to checkbox groups for clarity
  * Style the checkbox UI
  * Use buttons in modal footer for form submission (instead of links)
  * Attach labels to file download links

## How to test?

* Review this page (for instance) http://gigadb.gigasciencejournal.com:9170/dataset/100020 and run AXE audit (https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd)
  * Confirm that heading levels are not skipped
  * Confirm that tab labels do not display color contrast errors
  * Confirm that there are no errors about inputs without associated label within the table settings modal
  * Open table settings modal for samples and files, use tab key to navigate and confirm that checkboxes have visible focus, and confirm that modal buttons are <buttons> and not <a>
  * Confirm that download links have a defined aria-label

## How have functionalities been implemented?

* The layout and styles of the checkboxes in the table settings modal relied on styling a :before element and have been updated while keeping the gigadb color, by using the accent-color property. This allows to keep the native inputs visible, which is expected by some acceptance tests
* The checkboxes were wrapped in fieldset to indicate they are a related group

## Any issues with implementation?

One of the acceptance tests keeps failing. The reason escapes me as the changes in this PR seem unrelated to the page being tested by this test. Is anyone able to shed some light on why this test might be failing?

This is the command I run for the failing test:

`docker-compose run --rm test vendor/behat/behat/bin/behat --profile no-secrets -v --stop-on-failure features/dataset-admin.feature:19`

Based on what I see in the test spec, the Save submit button is not detected in http://gigadb.gigasciencejournal.com:9170/adminDataset/update/id/210 and I gather that a link is expected instead

The commit that caused this issue is https://github.com/gigascience/gigadb-website/pull/1426/commits/784cc73ff9201d6fe80cd88cf6b68529334681d0

I reverted the `<button>` in protected/views/dataset/_files_setting.php to be a `<a>`, and all tests pass. Note that semantically, a button is more suitable

However, the failing test and the change seem unrelated to me as the test seems to be testing a different page. Any idea for what reason this might be happening?

## Any changes to automated tests?

None

## Any changes to documentation?

None

## Any technical debt repayment?

None

## Any improvements to CI/CD pipeline?

None
